mulog-prometheus/src/com/brunobonacci/mulog/publishers/prometheus/collector.clj

```diff
@@ -0,0 +1,160 @@
+
+(extend-type Histogram$Child
+  ObserveValue
+  (observe-value [t value] (.observe t value)))
+
+(extend-type Summary$Child
+  ObserveValue
+  (observe-value [t value] (.observe t value)))
+
+
+
+
+(defn- simple-collector-builder
+  [builder-constructor
+   {:metric/keys [name namespace description label-keys]}]
+  (-> ^SimpleCollector$Builder (builder-constructor)
+    (.name name)
+    (.namespace namespace)
+    (.help description)
+    (.labelNames label-keys)
+    (.create)))
+
+
+
+(defmulti create-collection
+  "Dispatches on the `:metric/type`.
+  Returns a `SimpleCollector`"
+  (fn [metric] (:metric/type metric)))
+
+(defmethod create-collection :counter
+  [metric]
+  (simple-collector-builder #(Counter/build) metric))
+
+(defmethod create-collection :gauge
+  [metric]
+  (simple-collector-builder #(Gauge/build) metric))
+
+(defmethod create-collection :histogram
+  [{:metric/keys [buckets] :as metric}]
+  (simple-collector-builder #(cond-> (Histogram/build)
+                               (seq buckets) (.buckets buckets))
+    metric))
+
+(declare summary-builder)
+(defmethod create-collection :summary
+  [{:metric/keys [quantiles max-age-seconds age-buckets] :as metric}]
+  (simple-collector-builder #(summary-builder quantiles max-age-seconds age-buckets) metric))
+
+
+(defn add-quantile [^Summary$Builder builder [quantile error]]
+  (.quantile builder quantile error))
+
+(defn summary-builder
+  [quantiles max-age-seconds age-buckets]
+  (cond-> (Summary/build)
+    age-buckets     (.ageBuckets age-buckets)
+    max-age-seconds (.maxAgeSeconds max-age-seconds)
+    :always         ((partial reduce add-quantile) (or (seq quantiles) summary-quantiles-default))))
+
+
+
+
+(defmulti record-collection
+  "Receives a `[metric collection]`.
+  Dispatches on the `:metric/type`.
+  This will cause the collection to realise the events value.
+  Returns `[metric collection]`"
+  (fn [[metric collection]] (:metric/type metric)))
+
+(defmethod record-collection :counter
+  [[{:metric/keys [label-values]} collection :as met-col]]
+  (increment (child-with-labels collection label-values))
+  met-col)
+
+(defmethod record-collection :gauge
+  [[{:metric/keys [label-values value]} collection :as met-col]]
+  (set-value (child-with-labels collection label-values) value)
+  met-col)
+
+(defmethod record-collection :histogram
+  [[{:metric/keys [label-values value]} collection :as met-col]]
+  (observe-value (child-with-labels collection label-values) value)
+  met-col)
+
+(defmethod record-collection :summary
+  [[{:metric/keys [label-values value]} collection :as met-col]]
+  (observe-value (child-with-labels collection label-values) value)
+  met-col)
+
+
+
+(defn cleanup-metrics
+  [metrics]
+  {:pre [(s/every #(s/valid? ::ms.metric %))]}
+  (->> metrics
+    (remove nil?)
+    (map (fn [{:metric/keys [type value namespace name description labels buckets] :as m}]
+           (let [new-name (str name "_" (get metric-suffix type))]
```

Comment:
At this stage we should already have the final metric's names.  if the user uses `:transform-metrics` to provide metrics that he wants to be published I don't think we should change the names afterwards

Reply:
Agreed, corrected as per the metric type capturing correction